src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php

```diff
@@ -359,4 +370,51 @@ private function getContainerEnvVars(ContainerBuilder $container): array
 
         return array_values($envs);
     }
+
+    protected function getUsagesForDefinition(Definition $definition): array
+    {
+        $class = $definition->getClass();
+
+        return null !== $class ? $this->getUsages($class) : [];
+    }
+
+    private function getUsages(string $class): array
+    {
+        return $this->usages[$class] ?? [];
+    }
+
+    private function buildUsageMapForContainer(ContainerBuilder $builder)
+    {
+        foreach ($builder->getDefinitions() as $id => $definition) {
```

Comment:
Instead of iterating on each defintion I wonder if we couldn't leverage either:
- the `ServiceReferenceGraph` from the `Compiler` while provide nodes with `InEdges` and `OutEdges`....
- use the PhpDumper that already check this for CirculaReference detection and could provide a methods that returns the usage graph.

Reply:
I’ve replaced the usage map with ServiceReferenceGraph, but this results in different usages. This is probably caused by the way the graph is being built as it accounts for more use cases than I did previously. Could you check if this is correct?